### PR TITLE
feat(NgComponentOutlet): add NgComponentOutlet directive

### LINKED ISSUE #11168: [feature] NgComponentOutlet
**I'm submitting a ...**  (check one with "x")

```
[ ] bug report => search github for a similar issue or PR before submitting
[ x ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
None

**Expected/desired behavior**
A directive that is able to create components dynamically, the same way as `NgTemplateOutlet` created embedded view dynamically.
#### Pros
- Remove the boilerplate in a component. Developer does not need to query (`ViewChild`) and manually create the component. (`ComponentFactoryResolver`, `ViewContainerRef`)
- Expressed in the template.
- Simple
- Aligned with NgTemplateOutlet, same idea same concept
#### Cons
- Host attribute bindings are not expressed in a template. Not much of a con, as its expected and `ComponentRef` can be used to set values on the instance...
- Getting a reference to the returned `ComponentRef` is via callback only.
- more ?

**Directive**

``` ts
@Directive({
  selector: '[swapCmp]'
})
export class SwapComponentDirective {
  private component: any;

  constructor(private cfr: ComponentFactoryResolver,
              private vcRef: ViewContainerRef,
              private tRef: TemplateRef<Object>) {
  }

  @Input() swapCmpBindings: ResolvedReflectiveProvider[];
  @Input() swapCmpInjector: Injector;
  @Input() swapCmpProjectables: any[][];

  @Output() onCreate: EventEmitter<ComponentRef<any>> = new EventEmitter<ComponentRef<any>>(false);

  @Input() set swapCmp(component: any) {
    this.component = component;
    this.vcRef.clear();
    if (this.component) {
      let injector = this.swapCmpInjector || this.vcRef.parentInjector;

      if (Array.isArray(this.swapCmpBindings) && this.swapCmpBindings.length > 0) {
        injector = ReflectiveInjector.fromResolvedProviders(this.swapCmpBindings, injector);
      }

      const cmpRef = this.vcRef.createComponent(
        this.cfr.resolveComponentFactory(component),
        this.vcRef.length,
        injector,
        this.swapCmpProjectables
      );

      cmpRef.changeDetectorRef.detectChanges();

      this.onCreate.emit(cmpRef);
    }
  }
}
```

```
 <template [swapCmp]="ctx.component" [swapCmpBindings]="ctx.bindings"   [swapCmpInjector]="ctx.injector" [swapCmpProjectables]="ctx.projectableNodes"></template>
```

> The name here is SwapComponent but it just as an example, took it from my [code base](https://github.com/shlomiassaf/angular2-modal/blob/master/src/components/angular2-modal/components/swap-component.directive.ts)
- A Developer provides a component, changing the component will re-create.
- Injector is optional, if not supplied the current injector is used.
- Bindings are used to allow additional providers to the injector (something like `viewProviders`)
- In case `ng-content` is needed the developer can supply projectableNodes.

What do you think?

### PR DESCRIPTION
**Pull request by mhevery**
feat(NgComponentOutlet): add NgComponentOutlet directive

Add NgComponentOutlet directive that can be used to dynamically create 
host views from a supplied component.

Closes #11168
Takes over PR #11235

### CODE REVIEW

**Comment by shlomiassaf** [comment]
@mhevery I have pushed the updated code to my branch.

Tests are passing.

It includes a rebase +  all the requested changed except the namings.

Commits should be squashed :)

There are some lines of codes commented out in the tests since **Output** is not supported in * syntax.

I think you should use it as a baseline

https://github.com/shlomiassaf/angular/tree/ng-component-outlet

**Comment by mhevery** [comment]
@shlomiassaf thanks, i am also making sure that `ngTemplateOutlet` and `ngComponentOutlet` are consistent and adding documentation. At this point I think the two have diverged too much to be able to reuse your branch. Sorry.

**Comment by mhevery** [comment]
@vicb ready for review.

**Comment by MaximCrabbe** [comment]
So how long does it take on average for a PR like this to be reviewed? (Just would like to know, not 'pushing'). This NgComponentOutlet seems usefull!

**Comment by shlomiassaf** [comment]
Holiday season I guess (:

File: modules/@angular/common/src/directives/ng_template_outlet.ts [hunk 1]
```diff
@@ -15,42 +15,47 @@ import {Directive, EmbeddedViewRef, Input, OnChanges, SimpleChanges, TemplateRef
...
  * @experimental
  */
 @Directive({selector: '[ngTemplateOutlet]'})
 export class NgTemplateOutlet implements OnChanges {
   private _viewRef: EmbeddedViewRef<any>;
-  private _context: Object;
-  private _templateRef: TemplateRef<any>;
 
-  constructor(private _viewContainerRef: ViewContainerRef) {}
+  @Input() public ngTemplateOutletContext: Object;
```
Comment (vicb) [inline modules/@angular/common/src/directives/ng_template_outlet.ts]: What about `@Input('ngTemplateOutletContext') public context: Object;`

File: modules/@angular/common/src/directives/ng_template_outlet.ts [hunk 2]
```diff
@@ -15,42 +15,47 @@ import {Directive, EmbeddedViewRef, Input, OnChanges, SimpleChanges, TemplateRef
...
 
-  @Input()
-  set ngOutletContext(context: Object) { this._context = context; }
+  @Input() public ngTemplateOutlet: TemplateRef<any>;
```
Comment (vicb) [inline modules/@angular/common/src/directives/ng_template_outlet.ts]: What about `@Input('ngTemplateOutlet') public template: TemplateRef<any>;`

**Comment by vicb** [comment]
LGTM, please rebase and consider the minor comments.

**Comment by mhevery** [comment]
Rebased

File: modules/@angular/common/src/directives/ng_template_outlet.ts (hunk 1 above)
Comment (pkozlowski-opensource) [inline modules/@angular/common/src/directives/ng_template_outlet.ts]: IMO it is better as-is

File: modules/@angular/common/src/directives/ng_template_outlet.ts (hunk 2 above)
Comment (pkozlowski-opensource) [inline modules/@angular/common/src/directives/ng_template_outlet.ts]: IMO it is better as-is

File: modules/@angular/common/src/directives/ng_component_outlet.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  @Input() ngComponentOutletContent: any[][];
+
+  componentRef: ComponentRef<any>;
+
+  constructor(
+      private _cmpFactoryResolver: ComponentFactoryResolver,
+      private _viewContainerRef: ViewContainerRef) {}
+
+  ngOnChanges() {
+    this._viewContainerRef.clear();
```
Comment (pkozlowski-opensource) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: Could we change it so only views we've created are destroyed here? In theory we could have multiple directives sitting on the same `<template>` element, they could be creating their own views and this logic would mess up views created by other directives, no?

This is what is done in the template outlet: https://github.com/angular/angular/blob/35f9a1c2cb743d050e000cb8865a70240ce3ed6d/modules/%40angular/common/src/directives/ng_template_outlet.ts#L48-L50

Comment (mhevery) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: good point. Fixed

File: modules/@angular/examples/common/ngComponentOutlet/ts/module.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+class NgTemplateOutletSimpleExample {
+  // This field is necessary to expose HelloWorld to the template.
+  HelloWorld = HelloWorld;
+}
+// #enddocregion
+
+// #docregion CompleteExample
+@Injectable()
+class Greeter {
+  suffix = '!'
```
Comment (lgleim) [inline modules/@angular/examples/common/ngComponentOutlet/ts/module.ts]: [32, 15]: Missing semicolon.

File: modules/@angular/common/src/directives/ng_component_outlet.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+ */
+
+import {ComponentFactoryResolver, ComponentRef, Directive, Injector, Input, OnChanges, Provider, SimpleChanges, Type, ViewContainerRef} from '@angular/core';
+
+
+/**
+ * Instantiates a single {@link Component} type and inserts its Host View into current View.
+ * `NgComponentOutlet` provides a declarative approach for dynamic component creation.
+ *
+ * `NgComponentOutlet` requires a component type, if a falsy value is set the view will clear and
```
Comment (lgleim) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: Should probably be `if a false value` instead of `if a falsy value`.

**Review by lgleim — CHANGES_REQUESTED** [review]

**Comment by matsko** [comment]
@mhevery please fix CI issues.

**Comment by matsko** [comment]
Landed as https://github.com/angular/angular/commit/8578682dcfccc2ccccb68df37c3ebc738f98bf0e

**Comment by DzmitryShylovich** [comment]
beta 4

**Pull request closed without merging**